cmd/controller/certificates.go

```diff
@@ -217,23 +218,28 @@ func (c *Certificates) Init(ctx context.Context) error {
 	return eg.Wait()
 }
 
-func detectLocalIPs() ([]string, error) {
-	var localIPs []string
-	addrs, err := net.LookupIP("localhost")
+func detectLocalIPs(ctx context.Context) ([]string, error) {
+	resolver := net.DefaultResolver
+
+	addrs, err := resolver.LookupIPAddr(ctx, "localhost")
 	if err != nil {
 		return nil, err
 	}
 
 	if hostname, err := os.Hostname(); err == nil {
-		hostnameAddrs, err := net.LookupIP(hostname)
+		hostnameAddrs, err := resolver.LookupIPAddr(ctx, hostname)
 		if err == nil {
 			addrs = append(addrs, hostnameAddrs...)
+		} else if errors.Is(err, ctx.Err()) {
+			return nil, err
 		}
 	}
 
+	var localIPs []string
 	for _, addr := range addrs {
-		if addr.To4() != nil {
-			localIPs = append(localIPs, addr.String())
+		ip := addr.IP
+		if ip.To4() != nil || ip.To16() != nil {
```

Comment:
Isn't this the only change needed? net.LookupIP calls DefaultResolver.LookupIPAddr, so I don't think the other changes are doing anything at all.

Anyway the code looks correct to me, so LGTM

Reply:
Explanation in the PR description:

> Moreover, forward the context to detectLocalIPs(), so that the resolution may be cancelled.

So it's now using the context and DNS lookups can be cancelled in case they're hanging due to some misconfiguration.

Reply:
Ah good point.